tests/ffi/tests.cc

```diff
@@ -860,6 +864,12 @@ extern "C" const char *cxx_run_test() noexcept {
   rust::String utf16_rstring = utf16_literal;
   ASSERT(utf8_rstring == utf16_rstring);
 
+  const char *bad_utf8_literal = "test\x80";
+  const char16_t *bad_utf16_literal = (const char16_t *) "t\0e\0s\0t\0\x1E\xDD\0";
```

Comment:
I believe this is UB because the `const char *` that is getting casted is not necessary sufficiently aligned. Every `char16_t *` is required to be 2-byte aligned in C++.

Pointer conversion is defined in terms of conversion from `cv void*` like this in [expr.reinterpret.cast]:

> An object pointer can be explicitly converted to an object pointer of a different type. When a prvalue `v` of object pointer type is converted to the object pointer type "pointer to cv T", the result is `static_cast<cv T*>(static_cast<cv void*>(v))`.

and the meaning of conversion from `cv void*` is defined in [expr.static.cast]:

> A prvalue of type "pointer to cv1 void" can be converted to a prvalue of type "pointer to cv2 T" .... If the original pointer value represents the address `A` of a byte in memory and `A` does not satisfy the alignment requirement of `T`, then the resulting pointer value is unspecified.

Reply:
Ah, fair enough. I figured out how to get an invalid UTF-16 sequence in a `u` literal (`\x` can dump whatever crap you want into the literal!) so the new approach both avoids the UB and is more readable.

Reply:
Thanks, looks good. This also fixes the problem with endianness. `"t\0e\0s\0t\0\x1E\xDD\0"` would only pass the assertion of starting with "test" on little endian systems. On big endian it starts with "琀攀猀琀".